### ISSUE TITLE
fyi, wc_get_product was called incorrectly

### ISSUE THREAD

**Issue by TerrePorter** [woocommerce/woocommerce#24875]
**Describe the bug**
wc_get_product was called incorrectly. wc_get_product should not be called before the woocommerce_init action.

WooCommerce Subscriptions  by Automattic – 2.6.2
WooCommerce by Automattic – 3.7.1

This is more of a FYI as it does not break anything. This error occurs when there are subscription products in the cart, and you do a Ajax call.

I was having problems with a Ajax request related to subscription products. While digging in the error log I noticed an oddity.

[21-Oct-2019 14:20:43 UTC] wc_get_product was called incorrectly. wc_get_product should not be called before the woocommerce_init action. This message was added in version 2.5.

I traced it thinking it might be the problem in my code. But I found that it was actually a caused in WooCommerce:Init https://github.com/woocommerce/woocommerce/blob/2365525dcbe66cdc5137c6447631a538895bfe67/includes/class-woocommerce.php#L554

```
// Classes/actions loaded for the frontend and for ajax requests.
if ( $this->is_request( 'frontend' ) ) {
   wc_load_cart();
}

$this->load_webhooks();

// Init action.
do_action( 'woocommerce_init' );
```

The wc_load_cart function tries to load the cart products and ended up triggering the error in the wc_attribute_label when trying to get the product object because the product function now complains if it is used before the woocommerce_init action has been done. 
(https://github.com/woocommerce/woocommerce/blob/2365525dcbe66cdc5137c6447631a538895bfe67/includes/wc-attribute-functions.php#L193)

A quick fix would be to just move the wc_load_cart function after the woocommerce_init action.

```
$this->load_webhooks();

// Init action.
do_action( 'woocommerce_init' );

// Classes/actions loaded for the frontend and for ajax requests.
if ( $this->is_request( 'frontend' ) ) {
   wc_load_cart();
}
```

Here is the full log entry.
```
[21-Oct-2019 14:20:43 UTC] wc_get_product was called incorrectly. wc_get_product should not be called before the woocommerce_init action. 
Backtrace: 
 require_once('wp-load.php'),
 require_once('wp-config.php'),
 require_once('wp-settings.php'),
 do_action('init'),
 WP_Hook->do_action,
 WP_Hook->apply_filters,
 WooCommerce->init,
 wc_load_cart,
 WooCommerce->initialize_cart,
 WC_Customer->__construct,
 WC_Data_Store->read,
 WC_Customer_Data_Store->read,
 WC_Data->read_meta_data,
 maybe_unserialize,
 unserialize,
 WC_Data->__wakeup,
 WC_Product_Subscription_Variation->__construct,
 WC_Product_Variation->__construct,
 WC_Product_Simple->__construct,
 WC_Product->__construct,
 WC_Data_Store->read,
 WC_Product_Variation_Data_Store_CPT->read,
 WC_Product_Variation_Data_Store_CPT->generate_attribute_summary,
 wc_get_formatted_variation,
 wc_attribute_label,
 wc_get_product,
 wc_doing_it_wrong. This message was added in version 2.5.
```

**To Reproduce**
My simple test code, with a variant subscription product in the shopping cart.
```
public function handleAjaxTest()
{
    echo "handleAjaxTest function";
}
add_action('wp_ajax_handle_ajax_test', 'handleAjaxTest');
```

**Comment by claudiosanches** [comment]
This is not an issue on WooCommerce, also the this message is from WooCommerce saying that some plugin or theme is incorrectly using `wc_get_product()` function.

> wc_get_product was called incorrectly. wc_get_product should not be called before the woocommerce_init action. This message was added in version 2.5.

The code is here:

https://github.com/woocommerce/woocommerce/blob/master/includes/wc-product-functions.php#L46-L65

So maybe is a problem in Subscriptions or how you register things for ajax.

**Comment by TerrePorter** [comment]
@claudiosanches  Wow, …

“This is not an issue on WooCommerce, also the this message is from WooCommerce saying that some plugin or theme is incorrectly using wc_get_product() function.”

I did consider that it was a plugin or some other code before posting.

But guess what, yesterday I found this button on my IDE, that does this thing called ‘live code debugging’. So, I clicked it and took the time to trace the code execution path, to my surprise it followed the “generated error message log entry” (backtrace extracted, I am assuming). Even that log shows the execution path, I don’t see any jumps out to other plugins.

But wait, this will crack you up even more. If I move those three lines of code that try to load the cart before the init done hook, the error magically goes way. Of course, it’s not logical at all that the process of loading a cart might actually load the product class at some point. 

But in all seriousness, I do want to really thank you so much for pointing out the error message, ‘wc_get_product was called incorrectly’, is located in the wc_get_product function. I would never have thought to look there in a million years. I feel so enlightened now. It was so helpful, since I have only been programming for a few days now and this is all so new to me.

Beyond the sarcasm, you clearly didn’t read the post or even consider the function execution process. I was going to break it down line by line for you, but I think I am just going to put this down as a really lame brush off and go on. Besides, it is more of a cosmetic warning than something that breaks anything anyways.
